File: src/main/java/com/dumbdogdiner/stickycommands/commands/KillCommand.java
package com.dumbdogdiner.stickycommands.commands;

import com.dumbdogdiner.stickyapi.common.translation.LocaleProvider;
import com.dumbdogdiner.stickycommands.StickyCommands;
import com.dumbdogdiner.stickycommands.utils.Constants;
import dev.jorel.commandapi.CommandAPI;
import dev.jorel.commandapi.annotations.Command;
import dev.jorel.commandapi.annotations.Default;
import dev.jorel.commandapi.annotations.Permission;
import dev.jorel.commandapi.annotations.arguments.AEntitySelectorArgument;
import dev.jorel.commandapi.arguments.EntitySelectorArgument.EntitySelector;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Entity;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Player;

import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * Kill an entity
 */
@Command("killE")
@Permission(Constants.Permissions.KILL_ENTITIES)
public class KillCommand {
    private static LocaleProvider localeProvider;

    public KillCommand(StickyCommands instance){
        localeProvider = instance.getLocaleProvider();
        CommandAPI.registerCommand(getClass());
    }

    @Default
    public static void killEntities(CommandSender sender, @AEntitySelectorArgument(EntitySelector.MANY_ENTITIES)Collection<Entity> entities){
        Map<String, Integer> entityTypes = new HashMap<>();
        for(Entity e : entities) {
            if(!e.hasPermission(Constants.Permissions.KILL_IMMUNE)) {
                entityTypes.put(e.getName(), entityTypes.getOrDefault(e.getName(), 0) + 1);
                // Not sure if this will work
                e.sendMessage(localeProvider.translate("kill.you-were-killed", localeProvider.newVariables()));
                if (e instanceof LivingEntity)
                    ((LivingEntity) e).setHealth(0);
                else
                    e.remove();
            }
        }

        for(Map.Entry<String, Integer> entity : entityTypes.entrySet()){
            var vars = localeProvider.newVariables();
            vars.put("ENTITY", entity.getKey());
            vars.put("QUANTITY", String.valueOf(entity.getValue()));
            sender.sendMessage(localeProvider.translate("kill.you-killed-entities", vars));
        }
    }

    @Default
    public static void killEntity(CommandSender sender, @AEntitySelectorArgument(EntitySelector.ONE_ENTITY) Entity entity){
        killEntities(sender, Collections.singleton(entity));
    }
}

File: bukkit/src/main/kotlin/commands/KillCommand.kt
/*
 * Copyright (c) 2021 DumbDogDiner <dumbdogdiner.com>. All rights reserved.
 * Licensed under the MIT license, see LICENSE for more information.
 */
package com.dumbdogdiner.stickycommands.commands

import com.dumbdogdiner.stickyapi.bukkit.util.SoundUtil
import com.dumbdogdiner.stickyapi.common.util.NotificationType
import com.dumbdogdiner.stickycommands.util.Constants
import com.dumbdogdiner.stickycommands.util.Variables
import dev.jorel.commandapi.CommandAPI
import dev.jorel.commandapi.arguments.EntitySelectorArgument
import dev.jorel.commandapi.executors.CommandExecutor
import dev.jorel.commandapi.executors.PlayerCommandExecutor
import org.bukkit.entity.Entity
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player

object killCommand {
    fun register() {
        CommandAPI.unregister("kill") // unregister minecraft's /kill command
        killCommandWithArgument.register()
        killCommandWithoutArgument.register()
    }
}

private val killCommandWithArgument = commandStub("kill", Constants.Permissions.KILL)
    .withArguments(EntitySelectorArgument("entities", EntitySelectorArgument.EntitySelector.MANY_ENTITIES))
    .executes(CommandExecutor { sender, args ->
        val entities = args[0] as Collection<Entity>
        val vars = Variables().withSender(sender).get()
        if (entities.size == 1 && entities.firstOrNull() != null && entities.first() is Player) {
            val target = entities.first() as Player
            vars.putAll(Variables().withPlayer(target, true).get())
            when (true) {
                (target != sender && target.hasPermission(Constants.Permissions.KILL_IMMUNE)) -> {
                    sender.sendMessage(locale.translate(Constants.LanguagePaths.KILL_IMMUNE, vars))
                    SoundUtil.send(sender, NotificationType.ERROR)
                    return@CommandExecutor
                }
                (target != sender) -> {
                    sender.sendMessage(locale.translate(Constants.LanguagePaths.YOU_KILLED, vars))
                    target.sendMessage(locale.translate(Constants.LanguagePaths.YOU_WERE_KILLED, vars))
                }
                (target == sender) -> sender.sendMessage(locale.translate(Constants.LanguagePaths.SUICIDE, vars))
            }
            target.health = 0.0
        } else {
            entities.forEach loop@{
                if (it is LivingEntity) {
                    if (it.hasPermission(Constants.Permissions.KILL_IMMUNE)) return@loop
                    it.health = 0.0
                    it.sendMessage(locale.translate(Constants.LanguagePaths.YOU_WERE_KILLED, vars))
                } else {
                    it.remove()
                }
            }
            vars["amount"] = entities.size.toString()
            sender.sendMessage(locale.translate(Constants.LanguagePaths.YOU_KILLED_ENTITIES, vars))
    }

    SoundUtil.send(sender, NotificationType.SUCCESS)
    })

private val killCommandWithoutArgument = commandStub("kill", Constants.Permissions.KILL)
    .executesPlayer(PlayerCommandExecutor { sender, _ ->
        sender.health = 0.0
        sender.sendMessage(locale.translate(Constants.LanguagePaths.SUICIDE, Variables().withPlayer(sender, false).get()))
        SoundUtil.sendSuccess(sender)
    })
